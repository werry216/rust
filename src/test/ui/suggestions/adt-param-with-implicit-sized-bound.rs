trait Trait {
    fn func1() -> Struct1<Self>; //~ ERROR E0277
    fn func2<'a>() -> Struct2<'a, Self>; //~ ERROR E0277
    fn func3() -> Struct3<Self>; //~ ERROR E0277
}

struct Struct1<T>{
    _t: std::marker::PhantomData<*const T>,
}
struct Struct2<'a, T>{
    _t: &'a T,
}
struct Struct3<T>{
    _t: T,
}

fn main() {}
