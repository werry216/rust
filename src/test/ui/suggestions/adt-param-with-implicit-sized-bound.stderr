error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> $DIR/adt-param-with-implicit-sized-bound.rs:2:19
   |
LL |     fn func1() -> Struct1<Self>;
   |                   ^^^^^^^^^^^^^ doesn't have a size known at compile-time
...
LL | struct Struct1<T>{
   |                - required by this bound in `Struct1`
   |
   = help: the trait `std::marker::Sized` is not implemented for `Self`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
help: consider further restricting `Self`
   |
LL |     fn func1() -> Struct1<Self> where Self: std::marker::Sized;
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider relaxing the implicit `Sized` restriction
   |
LL | struct Struct1<T: ?Sized>{
   |                 ^^^^^^^^

error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> $DIR/adt-param-with-implicit-sized-bound.rs:3:23
   |
LL |     fn func2<'a>() -> Struct2<'a, Self>;
   |                       ^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
...
LL | struct Struct2<'a, T>{
   |                    - required by this bound in `Struct2`
   |
   = help: the trait `std::marker::Sized` is not implemented for `Self`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
help: consider further restricting `Self`
   |
LL |     fn func2<'a>() -> Struct2<'a, Self> where Self: std::marker::Sized;
   |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider relaxing the implicit `Sized` restriction
   |
LL | struct Struct2<'a, T: ?Sized>{
   |                     ^^^^^^^^

error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> $DIR/adt-param-with-implicit-sized-bound.rs:4:19
   |
LL |     fn func3() -> Struct3<Self>;
   |                   ^^^^^^^^^^^^^ doesn't have a size known at compile-time
...
LL | struct Struct3<T>{
   |                - required by this bound in `Struct3`
   |
   = help: the trait `std::marker::Sized` is not implemented for `Self`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
help: consider further restricting `Self`
   |
LL |     fn func3() -> Struct3<Self> where Self: std::marker::Sized;
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
